### PR TITLE
Add RCCL and training warmup for HYBRID_SHARD stability

### PR DESCRIPTION
**Pull request by oyazdanb**
- Add warmup.py with RCCL communicator warmup, manual param sync, and training collectives warmup functions
- Integrate RCCL warmup in build_fsdp_model before FSDP init
- Add training warmup in main() before training loop
- Add TORCH_DIST_INIT_TIMEOUT support for dist.init_process_group and dist.new_group calls
- Add cuda.synchronize() before gradient clipping to prevent race
- Improve exception handling in finally block
- Update set_env_variables.sh with DOCKER_ENV_VARS array for automatic env propagation to Docker containers
- Remove duplicate logging in local_launch.sh (logs now only in logs/ directory via master_launch.sh)
- Add warmup config params to shampoo_opt_multi_node.yaml

### CODE REVIEW

File: src/aorta/training/fsdp_trainer.py [hunk 1]
```diff
@@ -845,6 +919,33 @@ def main(args: Optional[argparse.Namespace] = None, *, enable_rocm_metrics: bool
 
     profiler = StreamProfiler(env["device"])
 
+    # Training warmup: run a few forward/backward/optimizer steps to warm up collectives
+    if not fsdp_cfg.skip_training_warmup and fsdp_cfg.training_warmup_steps > 0:
```
[the remark this answers is not in the document]
Comment (amd-vivekag) [reply src/aorta/training/fsdp_trainer.py:923]: @oyazdanb please take a look at this review comment. I'm not sure if the code design is as you intended.

File: src/aorta/utils/warmup.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+    Args:
+        shard_group: Intra-node shard process group (may be None)
+        replicate_group: Inter-node replicate process group (may be None)
+        device: GPU device to use for warmup tensors
+        num_warmup_ops: Number of warmup operations to perform (default: 5)
+    """
+    rank = dist.get_rank()
+    # Use a larger tensor for more thorough warmup
+    warmup_tensor = torch.ones(8192, device=device, dtype=torch.float32)
```
Comment (amd-vivekag) [inline src/aorta/utils/warmup.py:49]: instead of magic numbers, use code something like:
```
# At module level
WARMUP_TENSOR_SIZE = 8192  # 32KB per GPU (float32)
# Chosen to exercise memory channels without excessive overhead

# In function:
warmup_tensor = torch.ones(WARMUP_TENSOR_SIZE, device=device, dtype=torch.float32)
```

File: scripts/multi_node/set_env_variables.sh [hunk 3]
```diff
@@ -1,42 +1,133 @@
...
+# Network Interface for MI350X cluster
+# -----------------------------------------------------------------------------
+export NCCL_SOCKET_IFNAME=enp49s0f0np0,fenic0
 
+# -----------------------------------------------------------------------------
+# Timeout and Error Handling
+# -----------------------------------------------------------------------------
+export NCCL_TIMEOUT_MS=12000                 # 12 second timeout (legacy, not used by PyTorch)
+export NCCL_TIMEOUT=100                      # 100 second timeout - first backward can be slow due to JIT/init
+export TORCH_DIST_INIT_TIMEOUT=150           # Match collective timeout for consistency
```
Comment (amd-vivekag) [inline scripts/multi_node/set_env_variables.sh:57]: timeout `NCCL_TIMEOUT` and `TORCH_DIST_INIT_TIMEOUT` are pointing to different values. Any reason why do we have different values and different environment variables for same thing? Which one is effective in our code?

File: src/aorta/utils/warmup.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            dist.all_reduce(warmup_tensor, group=replicate_group)
+
+        # Synchronize GPU and global barrier between iterations
+        torch.cuda.synchronize()
+        dist.barrier()
+
+    # Final synchronization with extra delay
+    torch.cuda.synchronize()
+    dist.barrier()
+    torch.cuda.synchronize()
```
Comment (amd-vivekag) [inline src/aorta/utils/warmup.py:117]: double calls of sync and barrier. Need to remove one set

File: src/aorta/utils/warmup.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+            optimizer.step()
+
+        # Synchronize all ranks after each warmup step
+        torch.cuda.synchronize()
+        dist.barrier()
+
+        log.debug("Warmup step %d complete (rank=%d, loss=%.4f)", warmup_step, rank, loss.item())
+
+    # Reset optimizer state after warmup to not affect actual training
+    optimizer.zero_grad(set_to_none=True)
```
Comment (amd-vivekag) [inline src/aorta/utils/warmup.py:254]: Issue: zero_grad() only clears gradients, not optimizer state (momentum, learning rate schedule, etc.). Optimizers like Adam/Shampoo maintain state that's updated during warmup steps.
Problems:
- Warmup steps consume learning rate schedule steps
- Momentum buffers are initialized with warmup gradients
- Shampoo preconditioner is updated with warmup data

There are 3 options:

1. save optimizer state before warm up and restore after warmup.

```
# After warmup, reset optimizer to initial state
optimizer_state_backup = optimizer.state_dict()
# ... run warmup ...
# Restore initial state (but this is tricky with FSDP)
```

2. Create fresh optimizer after warmup
3. Document that warmup affects optimizer state

File: src/aorta/utils/warmup.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    torch.cuda.synchronize()
+    dist.barrier()
+
+    log.info("Manual parameter synchronization complete (rank=%d, params=%d)", rank, param_count)
+
+
+def warmup_training_collectives(
+    model: nn.Module,
+    optimizer: torch.optim.Optimizer,
+    dataloader,
```
Comment (amd-vivekag) [inline src/aorta/utils/warmup.py:182]: missing type hints

**Review by amd-vivekag — CHANGES_REQUESTED** [review]
